### PR TITLE
[feature](nereids) Support row policy

### LINKED ISSUE #13622: add row level policy in Nereids

### PR DESCRIPTION
**Pull request by xiaojunjie**
# Proposed changes

Issue Number: close #13622 

## Problem summary

This pr supports row policy  in nereids, which  did two things:
1. 【new logical plan】add **LogicalCheckPolicy** before UnboundRelation in LogicalPlanBuilder.
eg: `select * from xxx;`
       -- LogicalProject
       ------ **LogicalCheckPolicy**
       ---------- UnboundRelation
2. 【new rule】In NereidsAnalyzer, turn **LogicalCheckPolicy** to LogicalFilter if row policy exist, otherwise remove it. 
If there are multiple row policies, they will be merged into one, according to the original method.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
5. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
6. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
7. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
8. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.65 seconds
 load time: 570 seconds
 storage size: 17154644692 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221101180921_clickbench_pr_37579.html

File: fe/fe-core/src/main/java/org/apache/doris/policy/Policy.java [hunk 1]
```diff
@@ -91,8 +94,18 @@ public static Policy fromCreateStmt(CreatePolicyStmt stmt) throws AnalysisExcept
                 UserIdentity userIdent = stmt.getUser();
                 userIdent.analyze(ConnectContext.get().getClusterName());
                 TableIf table = db.getTableOrAnalysisException(stmt.getTableName().getTbl());
+                Expression nereidsPredicate = null;
+                if (ConnectContext.get() != null
+                        && ConnectContext.get().getSessionVariable() != null
+                        && ConnectContext.get().getSessionVariable().isEnableNereidsPlanner()) {
```
Comment (stalary) [inline fe/fe-core/src/main/java/org/apache/doris/policy/Policy.java:100]: Only need judge `ConnectContext.get().getSessionVariable().isEnableNereidsPlanner()`?

Comment (xiaojunjie) [reply fe/fe-core/src/main/java/org/apache/doris/policy/Policy.java:100]: ctx could be null in some unit tests

**Review by yinzhijian — DISMISSED** [review]
LGMT

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java [hunk 2]
```diff
@@ -43,33 +43,41 @@
...
+    private boolean policyChecked;
+
     public UnboundRelation(List<String> nameParts) {
-        this(nameParts, Optional.empty(), Optional.empty());
+        this(nameParts, false);
+    }
+
+    public UnboundRelation(List<String> nameParts, boolean policyChecked) {
+        this(nameParts, policyChecked, Optional.empty(), Optional.empty());
     }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:54]: The policyChecked hacking the UnboundRelation, I can imagine that there will be lots of field need added.
How about generate a CheckPolicy in the LogicalPlanBuilder?
And then you can write a rule to generate a new filter to the logicalRelation, like this:

```java
logicalCheckPolicy(logicalRelation()).thenApply(ctx ->
    LogicalCheckPolicy checkPolicy = ctx.root;
    LogicalRelation relation = checkPolicy.child();
    Optional<String> filter = checkPolicy.getFilter(relation);
    if (filter.isEmpty()) {
        return relation; // no any policy, so we remove the LogicalCheckPolicy Plan
    }
    // has policy, we should remove LogicalCheckPolicy and add a filter to relation
    Expression filter = new NereidsParser().parseExpression(filter.get());
    return new LogicalFilter(filter, relation);
)
```

Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:54]: Use logicalRelation can let you not need to care whether the table is exists.

Comment (xiaojunjie) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:54]: So good an idea. I image that the sql of filter is obtained from wherePredicate.toSql() in RowPolicy. But I have difficulty in parsing it, such as "**k[#1] = 1**", due to function toSqlImpl of SlotRef。So I have to use NereidsParser to generate an additional expression of filter when creating a policy. Any good idea?

Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundRelation.java:54]: The legacy `Expr` is difficult to convert to Nereids's `Expression`.
Use NereidsParser to parse the originStmt in `RowPolicy` and get the `Expression` is more reasonable.
It is same as parse view sql in each sql which using the view, so you can keep easy.

**Comment by morrySnow** [comment]
plz rebase newest master code ~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            }
+        }
+        if (!andList.isEmpty() && !orList.isEmpty()) {
+            return new And(ExpressionUtils.and(andList), ExpressionUtils.or(orList));
+        } else if (andList.isEmpty()) {
+            return ExpressionUtils.or(orList);
+        } else if (orList.isEmpty()) {
+            return ExpressionUtils.and(andList);
+        } else {
+            return null;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java:174]: `Optional.of(null)`will throw exception

File: regression-test/suites/account_p0/test_nereids_row_policy.groovy [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    assertQueryResult 2
+
+    // (v = 1) and (k = 1)
+    createPolicy"policy2", "k = 1", "PERMISSIVE"
+    assertQueryResult 1
+
+   // (v = 1) and (k = 1 or k = 2)
+    createPolicy"policy3", "k = 2", "PERMISSIVE"
+    assertQueryResult 2
+
```
Comment (924060929) [inline regression-test/suites/account_p0/test_nereids_row_policy.groovy:90]: we should check row policy whether is valid in a view

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * get wherePredicate of policy for logicalRelation.
+     *
+     * @param logicalRelation include tableName and dbName
+     * @param connectContext include information about user and policy
+     */
+    public Optional<Expression> getFilter(LogicalRelation logicalRelation, ConnectContext connectContext) {
+        String dbName = !logicalRelation.getQualifier().isEmpty() ? logicalRelation.getQualifier().get(0) : null;
+        Database db = connectContext.getEnv().getInternalCatalog().getDb(dbName)
+                .orElseThrow(() -> new RuntimeException("Database [" + dbName + "] does not exist."));
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java:124]: The `LogicalTVFRelation` is `LogicalRelation` but should not throw exception.
I think we should provide some interfaces, like `CatalogRelation`, `InternalCatalogRelation`, `ExternalCatalogRelation`, then we check for `CatalogRelation`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 6]
```diff
@@ -510,6 +510,8 @@ private void eliminateFromGroupAndMoveToTargetGroup(Group fromGroup, Group targe
                     "TargetGroup should be ancestors of fromGroup, but fromGroup is root. Maybe a bug");
         }
 
+        moveParentExpressionsReference(fromGroup, targetGroup);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:513]: In the context, `fromGroup` is the group of `LogicalRelation`, and the `targetGroup` is the group of `LogicalRowPolicy`, `eliminateFromGroupAndMoveToTargetGroup` will move the logicalExpression(LogicalRowPolicy) to the targetGroup. And the origin logicalExpression in the targetGroup will be recycle until the `LogicalRelation`.

So I don't think we should `moveParentExpressionsReference(fromGroup, targetGroup)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * get wherePredicate of policy for logicalRelation.
+     *
+     * @param logicalRelation include tableName and dbName
+     * @param connectContext include information about user and policy
+     */
+    public Optional<Expression> getFilter(LogicalRelation logicalRelation, ConnectContext connectContext) {
+        if (!(logicalRelation instanceof CatalogRelation)) {
+            return Optional.empty();
+        }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalCheckPolicy.java:124]: good job

**Review by 924060929 — APPROVED** [review]

**Pull request merged**